### ISSUE TITLE
Linux x86_64 version of Kodi is a RAM hog

### ISSUE THREAD

**Issue by smp79** [xbmc/xbmc#15913]
### Describe the bug
Here is a clear and concise description of what the problem is:

After upgrading my LibreELEC system from Raspberry Pi 3 to Intel J4105 board with 2Gb of RAM I encountered an unexpected issue. It is impossible to use Kodi for a prolonged period of time without encountering an out-of-memory condition which results in OOM killer restarting Kodi.
The only addons I ever install and use are vdr-addon, pvr.vdr.vnsi and oscam. This is not an addon issue.
The easiest way for me to reproduce the issue is to tune to a high bitrate 4K TV channel and leave it playing for a few hours. During that time the memory usage of kodi.bin process will progressively grow. Stopping the playback does not make it release the used RAM. 
I tried to work around the issue by decreasing the MALLOC_MMAP_THRESHOLD value. I started with 131072 and went as low as 8192. It does not seem to make a huge difference - kodi.bin still grows and is eventually killed by OOM.

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

I expect kodi.bin RAM usage not to grow infinitely.

### Debuglog

The debuglog can be found here:
https://paste.kodi.tv/iminadayok.kodi

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [x] Linux
 - [ ] OSX
 - [ ] Raspberry-Pi
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: LibreELEC
 - Kodi version: 18.2-rc1 current master

**Comment by pkerling** [comment]
Hi, can you reproduce without PVR, i.e. by playing a "normal" video file?

**Comment by smp79** [comment]
I see kodi.bin grow at around 1% per hour when playing a high bitrate 1080p video file.

**Comment by olafhering** [comment]
Run it via valgrind to detect the leak.

**Comment by smp79** [comment]
I'm not sure how to run it with valgrind. It runs extremely slow and totally unusable.
I tried
`valgrind --leak-check=full /usr/lib/kodi/kodi.bin &> /storage/valgrind.log`

**Comment by olafhering** [comment]
Usage of valgrind always requires time and patience. In another project I used this script:

````
valgrind.massif.kodi.sh
#!/bin/bash
set -ex
exe='/usr/lib64/kodi/kodi-x11'
sexe="${exe//\//_}"
vg="$(type -P valgrind)"
x="$(type -P ${exe})"
test -n "${vg}"
test -n "${x}"
mmap='no'
case "${mmap}" in
        yes) mode=mmap ; heap=yes ;;
        no) mode=malloc ; heap=no ;;
        *) exit 1
esac
exec \
        "${vg}" \
        --tool=massif \
        --log-file=/dev/shm/valgrind.massif.${mode}.${sexe}.%p.%n.txt \
        --heap=yes \
        --time-unit=B \
        --threshold=0.1 \
        --pages-as-heap="${heap}" \
        --depth=30 \
        --massif-out-file=/dev/shm/valgrind.massif.${mode}.${sexe}.%p.%n.out \
        "${x}" "$@"
````

Depending on the kind of leak, --time-unit and threshold values have to be adjusted.
On my (totally different) system the valgrind process takes about 720MB when playing a physical DVD.

I think the /dev/shm path must be adjusted to write to real storage and not waste precious RAM.

Monitor the system with top while kodi is running via valgrind. See how fast it leaks. Later provide the log and out files.

Thanks.

**Comment by smp79** [comment]
I did not use valgrind but I did more testing and the memory leak is definitely triggered when VAAPI hardware acceleration is used.
With VAAPI hardware acceleration disabled kodi.bin memory usage does not increase over time.

I'm now testing on a Celeron J4105 board with 4Gb of RAM. This is how it looks after a couple of days of running 1080p videos non-stop with VAAPI enabled:

![capture_001_23042019_005300](https://user-images.githubusercontent.com/28512968/56535010-7c1db480-6563-11e9-903e-a5a9d9d1643a.jpg)

**Comment by pkerling** [comment]
That narrows it down a lot, thanks for the investigation. Do you also observe this with Kodi 18.1? (There were changes in VAAPI for 18.2)

**Comment by smp79** [comment]
I tried an older (July 2018) Milhouse build and I don't see any difference. It is leaking memory when VAAPI is used just like the current master.

**Comment by pkerling** [comment]
Could confirm on master. It's not much, but it's there.

**Comment by smp79** [comment]
I'm testing a generic OpenGLES build of LibreELEC (with X11 removed). There is no improvement whatsoever regarding the VAAPI memory usage. I also tried Intel's media-driver instead of the default intel-vaapi-driver. No improvement as well.

**Comment by pkerling** [comment]
fyi I tested on AMD, so it's not the driver

**Comment by kuraikaze** [comment]
For what it's worth I'm seeing this with VDPAU as well (on Generic x86_64 LE 9.0.2/Kodi 18.2) with an old Nvidia GT640 GPU.
It's a little early to be sure, since I haven't had time to give it long like-for-like runs, but with VDPAU disabled it seems that the memory leak is much reduced, if not altogether gone - reduced enough that it's hard to detect against the normal background noise of accumulated memory cruft of a long uptime session.

Could this imply that it's an issue with *any* hardware acceleration handoff ?

I compiled a debug + valgrind build but, like smp79, I found running it against kodi.bin it to be so slow and the logs generated so huge that if there *was* some useful result it was hidden in all the noise.
I'll try again with the valgrind massif tool, instead of the default memcheck, sometime over the next few days and see if it produces a more useable output.

[https://forum.libreelec.tv/thread/17579-how-to-properly-diagnose-a-persistent-memory-leak/](url)

**Comment by smp79** [comment]
Another thing I noticed is that the VAAPI leak appear to be worse when there is video upscaling or downscaling is involved (e.g. 720p-->1080p or 4k-->1080p).

**Comment by mglae** [comment]
I've build a recent 18.3-rc with `-fsanitize=leak -fno-omit-frame-pointer`on my desktop. The log is at [http://ix.io/1Lm4](http://ix.io/1Lm4).

Using `addr2line` on the VAAPI::CDecoder::Decode() addresses result in [VAAPI.cpp:869](https://github.com/xbmc/xbmc/blob/7d3fee524634782e6ad5414773d51b63581d8b5a/xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.cpp#L869) and [VAAPI.cpp:874](https://github.com/xbmc/xbmc/blob/7d3fee524634782e6ad5414773d51b63581d8b5a/xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.cpp#L874). The leaking code is the same on VDPAU.

@smp79, @kuraikaze please check if there is improvement with this [patch](http://ix.io/1Lm5).

**Comment by smp79** [comment]
@mglae patch fixed the issue. Playing a DVB stream for 3+ hours now, Kodi memory usage does not increase. Thank you!

**Comment by fritsch** [comment]
@yol @FernetMenta the change looks sane to me. It's also the only place where the ptr can be freed. Do you agree?

**Comment by fritsch** [comment]
@mglae thx very much for digging that deep into it.

**Comment by mglae** [comment]
@fritsch it's on my todo list to check if the delete is needed [here](https://github.com/xbmc/xbmc/blob/7d3fee524634782e6ad5414773d51b63581d8b5a/xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.cpp#L1590).

**Comment by fritsch** [comment]
I think it is needed.

**Comment by pkerling** [comment]
> @yol @FernetMenta the change looks sane to me. It's also the only place where the ptr can be freed. Do you agree?

Not really I'm afraid. Even though the raw pointer usage does not make it very clear (I also was confused at first), there is a defined ownership hierarchy. The `CVaapiDecodedPicture` is owned by the `CPayloadWrap` which puts it into a `std::unique_ptr` here: https://github.com/xbmc/xbmc/blob/4398f98/xbmc/utils/ActorProtocol.h#L38
I.e., the decoded picture instance is destroyed together with the payload wrap. The `CPayloadWrap` is owned by the message sent to the message queue of the port.  It is in turn put into a `std::unique_ptr` here: https://github.com/xbmc/xbmc/blob/4398f988e1577495ab6a9ba1e43e6bd2ac65f465/xbmc/utils/ActorProtocol.cpp#L159
 -> it  is supposed to be destroyed together with the message

In my view, this means that these changes work around an issue with the ActorProtocol implementation. The root problem seems to be here: https://github.com/xbmc/xbmc/blob/4398f988e1577495ab6a9ba1e43e6bd2ac65f465/xbmc/utils/ActorProtocol.cpp#L31

I think this line should read `payloadObj.reset()` instead of `payloadObj.release()`, because `release()` will just give up ownership and not free any memory. Not using the returned pointer is almost guaranteed to be a memory leak. @FernetMenta This is your code and you know the intent best - do you agree or is it a design decision to not release the payload memory?

**Comment by fritsch** [comment]
I thought so first, but then I wondered as the receiver of the message uses the RAW ptr. Means starting at this point the data would vanish when the smart_ptr goes out of scope and the release was replaced with reset();

That's why I thought the "consumer" of the raw ptr should be the one deleting it. Let's see.

But let's wait on Fernetmenta.

**Comment by FernetMenta** [comment]
@pkerling you are right, should be reset. I think I was not careful enough with autocompletion.

**Comment by smp79** [comment]
I tested with payloadObj.reset() for a few hours, it also fixed the issue for me.

**Comment by smp79** [comment]
I noticed another, possibly related issue when testing payloadObj.reset() fix. When I pause the video - kodi.bin memory usage start to grow quite rapidly. If I disable VAAPI hardware acceleration the issue is gone.

**Comment by smp79** [comment]
> I noticed another, possibly related issue when testing payloadObj.reset() fix. When I pause the video - kodi.bin memory usage start to grow quite rapidly. If I disable VAAPI hardware acceleration the issue is gone.

Ok, the memory issue with the paused video is probably unrelated. It is specific to a GBM version of Kodi. I can't reproduce this with the x11 version. I will open a new problem report.

**Comment by candrews** [comment]
Did you mean to close this issue?

As far as I can tell, there hasn't been a commit fixing this issue or a PR suggesting one (presumably the release -> reset change). If I'm mistaken, can you please link to it?

**Comment by smp79** [comment]
Well, I guess I should keep it open until the fix is merged.

**Comment by pkerling** [comment]
I'm a little short on time at the moment so if anyone wants to PR go ahead
